### ISSUE TITLE
supporting iptables log limit feature

### ISSUE THREAD

**Issue by natieitan** [projectcalico/calico#2392]
During our migration to Calico security, we end the policy with **log** and allow rule in order to see which rules are still missing in the policy. Because Calico implements the rules with iptables the iptables logs all traffic within the **kernel syslog** facility. When high traffic occurs the log facility cannot handle all the messages therefor it suppresses messages that cause the iptables to crush.
In order to overcome this issue the iptables supports **limit** module as explained in the following link: [(https://making.pusher.com/per-ip-rate-limiting-with-iptables/)]

## Expected Behavior
We expect Calico to be able to accept a flag that can support the limit module interpreted to iptables command.

## Current Behavior
Not able to apply limits on iptables within Calico

## Your Environment
* Calico version - 3.3.2
* Orchestrator version - kubernetes 1.10
* Operating System and version: CentOS 7

**Comment by mhmxs** [comment]
@caseydavenport Limitation itself is a complex problem in iptables, it has many flags, behaviors, etc. So as i see there are two ways to fix this issue:
a) Solve it as simple as possible only for `LOG`, witch means if user provides limitation in the `NetworkPolicy` description, Calico extends the iptables command with a simple `--match limit --limit X/sec`. Nothing else.
b) Calico starts support limitations generally with `limit` and `hashlimit` with all the available flags of them.

Can i ask what is your advice?

**Comment by neiljerram** [comment]
@natieitan Can you provide more detail for "When high traffic occurs the log facility cannot handle all the messages therefor it suppresses messages that cause the iptables to crush"?  I'd kinda expect any log facility to be able to handle massive incoming load, and to provide rate limiting features (for output to files, for example) itself, so it's not obvious to me that there should also be a need for rate limiting in the source of the logs.

**Comment by neiljerram** [comment]
@mhmxs Can you expand on what you're suggesting for (b) ?  Does it mean rate limiting for iptables rules that are not `-j LOG` ?

**Comment by mhmxs** [comment]
@neiljerram Yes, instead for support it for just `LOG`, support it everywhere/everycase where it makes sense (in a generalized way), for example rate limiting. I know it is a huge work. Currently i'm just thinking in the scope and the goal of this issue.

**Comment by neiljerram** [comment]
@mhmxs Right, but I don't think there are any cases other than `LOG`.  If we have a rule with `DROP` or `ACCEPT`, for example, it really needs to apply to all packets :smile:

**Comment by neiljerram** [comment]
@caseydavenport Please correct if I'm wrong here, but I don't believe the need for this enhancement has really been explained.  Hence demoting to backlog, and will close if there's no more activity after a couple of weeks.

**Comment by caseydavenport** [comment]
@neiljerram I am not sure I've followed entirely - do we already set a limit on LOG rules today?

Agree a limit on non-LOG rules doesn't make sense - but it might be sensible to support a configurable / optional ratelimit on LOG rules?

**Comment by neiljerram** [comment]
@caseydavenport No, we don't already have a limit for LOG rules.  But I don't think the case has been made even for LOG rules.  We need more detail behind the original report where it says:

> When high traffic occurs the log facility cannot handle all the messages therefor it suppresses messages that cause the iptables to crush

If we can't get that, I think we should close this issue.

**Comment by natieitan** [comment]
@neiljerram For me to provide the logs of that issue I need to reproduce it, which isn't possible as these clusters hold our services.
IPtables write a lot of logs to /var/log/messages while trying to do so (if the load is high on that log rule) the kernel shut down the IPtables process on the K8S node which causes Calico-node to crush

**Comment by DavidDarville** [comment]
Another option would be to use the NFLOG target instead of the LOG one. Since NFLOG can handle a much higher rate of log messages.

**Comment by akhilesh2410** [comment]
> Another option would be to use the NFLOG target instead of the LOG one. Since NFLOG can handle a much higher rate of log messages.

Is NFLOG option is supported by Calico ? It seems calico only creates rules without NFLOG option. Is there a way we can capture packets with their exact verdict i.e. ALLOW or DENY based on the rule they hit ?

**Comment by CodeGlitcher** [comment]
Any news on this?
We have a log rule that logs a event if a request is blocked by the policy.
But in our test environment nodes are not that stable and are often restarted.
But because it takes a while for calico to catch up during this time we see a lot of logging from calico.
Sometimes we need to remove the global network policy to log blocked request, juist to get our cluster up and running.
